rts/motoko-rts/src/gc/mark_compact.rs

```diff
@@ -0,0 +1,312 @@
+//! Implements "threaded compaction" as described in The Garbage Collection Handbook section 3.3.
+
+use crate::bitmap::{alloc_bitmap, free_bitmap, get_bit, iter_bits, set_bit};
+use crate::closure_table::closure_table_loc;
+use crate::mark_stack::{self, alloc_mark_stack, free_mark_stack, pop_mark_stack};
+use crate::mem::memcpy_words;
+use crate::rts_trap_with;
+use crate::types::*;
+
+use super::{get_heap_base, get_static_roots, note_live_size, note_reclaimed, HP};
+
+#[no_mangle]
+unsafe extern "C" fn compacting_gc() {
+    let old_hp = HP;
+    let heap_base = get_heap_base();
+
+    mark_compact(heap_base, old_hp, get_static_roots(), closure_table_loc());
+
+    let reclaimed = old_hp - HP;
+    note_reclaimed(Bytes(reclaimed));
+
+    let new_live_size = HP - heap_base;
+    note_live_size(Bytes(new_live_size));
+}
+
+unsafe fn mark_compact(
+    heap_base: u32,
+    heap_end: u32,
+    static_roots: SkewedPtr,
+    closure_table_loc: *mut SkewedPtr,
+) {
+    let heap_size = Bytes(heap_end - heap_base);
+
+    alloc_bitmap(heap_size);
+    alloc_mark_stack();
+
+    mark_static_roots(static_roots, heap_base);
+
+    // TODO: We could skip the is_tagged_scalar, heap_base etc. checks
+    push_mark_stack(*closure_table_loc, heap_base);
+
+    mark_stack(heap_base);
+
+    thread_roots(static_roots, heap_base);
+    thread(closure_table_loc, heap_base);
+    update_fwd_refs(heap_base);
+    update_bwd_refs(heap_base);
+
+    free_bitmap();
+    free_mark_stack();
+}
+
+unsafe fn mark_static_roots(static_roots: SkewedPtr, heap_base: u32) {
+    let root_array = static_roots.as_array();
+
+    // Static objects are not in the dynamic heap so don't need marking.
+    for i in 0..root_array.len() {
+        let obj = root_array.get(i).unskew() as *mut Obj;
+        mark_fields(obj, heap_base);
+    }
+}
+
+unsafe fn push_mark_stack(obj: SkewedPtr, heap_base: u32) {
+    if obj.is_tagged_scalar() {
```

Comment:
would it make sense to do this at the call-sites to avoid the function call altogether?

Reply:
(not a requirement for now)

Reply:
I can try it, but this function is called in 11 places so the code will suffer a little bit.

Reply:
I wonder if LLVM can do "partial inlining" and only inline the early return path, effectively doing the transformation you suggest.

Reply:
To be honest, I have no expectation for what LLVM will do for you. Would a macro not let you keep the code more or less unchanged?

Reply:
Done as a part of 9af5279.